python/ray/data/_internal/execution/backpressure_policy.py

```diff
@@ -0,0 +1,100 @@
+import logging
+import os
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from ray.data._internal.execution.interfaces.physical_operator import (
+        PhysicalOperator,
+    )
+    from ray.data._internal.execution.streaming_executor_state import Topology
+
+logger = logging.getLogger(__name__)
+
+
+# TODO(hchen): Enable ConcurrencyCapBackpressurePolicy by default.
+DEFAULT_BACKPRESSURE_PLOCIES = []
+
+
+def get_backpressure_policies(topology: "Topology"):
+    return [policy(topology) for policy in DEFAULT_BACKPRESSURE_PLOCIES]
+
+
+class BackpressurePolicy(ABC):
+    """Interface for back pressure policies."""
+
+    @abstractmethod
+    def __init__(self, topology: "Topology"):
+        ...
+
+    @abstractmethod
+    def can_run(self, op: "PhysicalOperator") -> bool:
+        """Called when StreamingExecutor is about to select an operator to run.
+        Returns True if the operator can run, False otherwise.
+        """
+        ...
+
+
+class ConcurrencyCapBackpressurePolicy(BackpressurePolicy):
+    """A backpressure policy that caps the concurrency of each operator.
+
+    The concurrency cap limits the number of concurrently running tasks.
+    It will be set to an intial value, and will ramp up exponentially.
+
+    The concrete stategy is as follows:
+    - Each op is assigned an initial concurrency cap.
+    - An op can run new tasks if the number of running tasks is less than the cap.
+    - When the number of finished tasks reaches a threshold, the concurrency cap will
+      increase.
+    """
+
+    # Environment variable to configure this policy.
+    # The format is: "<init_cap>,<cap_multiply_threshold>,<cap_multiplier>"
+    CONFIG_ENV_VAR = "RAY_DATA_CONCURRENCY_CAP_CONFIG"
+
+    # The intial concurrency cap for each operator.
+    INIT_CAP = 4
+    # When the number of finished tasks reaches this threshold, the concurrency cap
+    # will be multiplied by the multiplier.
+    CAP_MULTIPLY_THRESHOLD = 0.5
+    # The multiplier to multiply the concurrency cap by.
+    CAP_MULTIPLIER = 2.0
```

Comment:
another way is to define these constant on the file level, so we don't need to parse environment variables. It's also maybe easier for advanced users to test out.

Reply:
The issue with constants is that if the executor doesn't run on the driver (e.g., on the SplitCoordinator actor), it's hard to change the configs. I've seen this issue for other configs that depend on constants. Do you know a good solution to bypass this issue?

Reply:
I think you'll need to put them into the DataContext.

Reply:
@stephanie-wang @c21 If we put them into DataContext, I'd like to to save them in a dict and add a key-value interface in DataContext. 
The reason is because this is a plugin and DataContext shouldn't need to know about the plugin configs. 
What do you think?

The API will be something like:
```python
data_context.set("concucrrency_cap_backpressure_policy.init_cap", 4)
```

Reply:
That sounds good to me for now while it's still experimental. Actually, could you prepend the name with "experimental" or something like that? Makes deprecation a bit smoother.